packages/cojson-transport-ws/src/index.ts

```diff
@@ -18,87 +18,54 @@ export type CreateWebSocketPeerOpts = {
     role: Peer["role"];
     expectPings?: boolean;
     batchingByDefault?: boolean;
+    onClose?: () => void;
 };
 
-export function createWebSocketPeer({
-    id,
-    websocket,
-    role,
-    expectPings = true,
-    batchingByDefault = true,
-}: CreateWebSocketPeerOpts): Peer {
-    const incoming = new cojsonInternals.Channel<
-        SyncMessage | DisconnectedError | PingTimeoutError
-    >();
-
-    websocket.addEventListener("close", function handleClose() {
-        incoming
-            .push("Disconnected")
-            .catch((e) =>
-                console.error("Error while pushing disconnect msg", e),
-            );
-    });
-
-    let pingTimeout: ReturnType<typeof setTimeout> | null = null;
-
-    let supportsBatching = batchingByDefault;
-
-    websocket.addEventListener("message", function handleIncomingMsg(event) {
-        const result = deserializeMessages(event.data as string);
-
-        if (!result.ok) {
-            console.error("Error while deserializing messages", event.data, result.error);
-            return;
+function createPingTimeoutListener(enabled: boolean, callback: () => void) {
+    if (!enabled) {
+        return {
+            reset() {},
+            clear() {}
         }
+    }
 
-        const { messages } = result;
-
-        if (!supportsBatching && messages.length > 1) {
-            // If more than one message is received, the other peer supports batching
-            supportsBatching = true;
-        }
+    let pingTimeout: ReturnType<typeof setTimeout> | null = null;
 
-        if (expectPings) {
+    return {
+        reset() {
             pingTimeout && clearTimeout(pingTimeout);
             pingTimeout = setTimeout(() => {
-                incoming
-                    .push("PingTimeout")
-                    .catch((e) =>
-                        console.error("Error while pushing ping timeout", e),
-                    );
+                callback();
             }, 10_000);
+        },
+        clear() {
+            pingTimeout && clearTimeout(pingTimeout);
         }
+    };
+}
 
-        for (const msg of messages) {
-            if (msg && "action" in msg) {
-                incoming
-                    .push(msg)
-                    .catch((e) =>
-                        console.error("Error while pushing incoming msg", e),
-                    );
-            }
-        }
-    });
-
-    const websocketOpen = new Promise<void>((resolve) => {
+function waitForWebSocketOpen(websocket: AnyWebSocket) {
+    return new Promise<void>((resolve) => {
         if (websocket.readyState === 1) {
             resolve();
         } else {
             websocket.addEventListener("open", resolve, { once: true });
         }
     });
+}
 
+function createOutgoingMessagesManager(websocket: AnyWebSocket, batchingByDefault: boolean) {
```

Comment:
Abstracted the outgoing messages sending to simplify the batching toggle management